include/proteus/JitFrontend.hpp

```diff
@@ -42,54 +44,36 @@ class JitModule {
     // Launch with type-safety.
     [[nodiscard]] auto launch(LaunchDims Grid, LaunchDims Block,
                               uint64_t ShmemBytes, void *Stream, ArgT... Args) {
-      // Create the type-safe tuple.
-      auto Tup = std::make_tuple(static_cast<ArgT>(Args)...);
-
-      // Create the ArrayRef<void*> pointing at each tuple element.
-      std::array<void *, sizeof...(ArgT)> Ptrs;
-      std::apply(
-          [&](auto &...Elts) {
-            size_t I = 0;
-            ((Ptrs[I++] = (void *)&Elts), ...);
-          },
-          Tup);
-
-      // Call launch through module.
-      // TODO: should it use the dispatcher directly?
-      return M.launch(F, Grid, Block, Ptrs, ShmemBytes, Stream);
-    }
+      // Pointers to the local parameter copies.
+      void *Ptrs[sizeof...(ArgT)] = {(void *)&Args...};
 
-    FuncBase *operator->() { return &F; }
-  };
+      if (!M.isCompiled())
+        M.compile();
 
-  TargetModelType parseTargetModel(StringRef Target) {
-    if (Target == "host" || Target == "native") {
-      return TargetModelType::HOST;
-    }
+      auto GetKernelFunc = [&]() {
+        // Get the kernel func pointer directly from the Func object if
+        // available.
+        if (auto KernelFunc = F.getCompiledFunc()) {
+          return KernelFunc;
+        }
 
-    if (Target == "cuda") {
-      return TargetModelType::CUDA;
-    }
+        // Get the kernel func pointer from the Dispatch and store it to the
+        // Func object to avoid cache lookups.
+        // TODO: Re-think caching and dispatchers.
+        auto KernelFunc = reinterpret_cast<decltype(F.getCompiledFunc())>(
+            M.Dispatch.getFunctionAddress(F.getName(), M.getObjectModuleRef()));
 
-    if (Target == "hip") {
-      return TargetModelType::HIP;
-    }
+        F.setCompiledFunc(KernelFunc);
```

Comment:
I don't love the idea that we are adding an unobservable third cache tier only to the frontend DSL.

Reply:
Me neither, I will open an Issue as a TODO, the problem is observability. The sane thing is to *indeed save* the pointer the handle because there is no point spending time in needless cache lookups, but we should make this observable.